### PR TITLE
Enable encryption with Sync

### LINKED ISSUE #4561: Enable encryption with Sync
We’re not using the encryption key for Synced Realm https://github.com/realm/realm-java/blob/fef417cb35a9920838a07f1ab93c8a84645b4d6f/realm/realm-library/src/main/cpp/io_realm_internal_SharedRealm.cpp#L134

this is because the ObjectStore has a different `realm_encryption_key` https://github.com/realm/realm-object-store/blob/3eada170f380174992b47b6023f375f3f372a9d2/src/sync/sync_config.hpp#L108 parameter that we need to set (as opposed to the non sync encryption param `encryption_key` https://github.com/realm/realm-object-store/blob/563f2e9993ae4f4a8ab6f524a1ba71e707ebe78f/src/shared_realm.hpp#L145)

https://github.com/realm/realm-object-store/pull/414

### PR DESCRIPTION
**Pull request by nhachicha**
fixes #4561

### CODE REVIEW

**Comment by nhachicha** [comment]
@realm/java

**Comment by zaki50** [comment]
What will happen if an app of old version is already using `encryptionKey()` and the app is updated to fixed version?
Existing Realm database file will be encrypted automatically? Just an exception will be thrown?

File: realm/realm-library/src/main/cpp/io_realm_internal_SharedRealm.cpp [hunk 1]
```diff
@@ -145,7 +145,11 @@ class JniConfigWrapper {
             user, realm_url, SyncSessionStopPolicy::Immediately, std::move(bind_handler), std::move(error_handler),
             nullptr, util::none, sync_client_validate_ssl, ssl_trust_certificate_path});
 
-
+        if (!m_config.encryption_key.empty()) {
```
Comment (beeender) [inline realm/realm-library/src/main/cpp/io_realm_internal_SharedRealm.cpp]: What about init the key before construct SyncConfig to make it less confusing?
```c++
        util::Optional<std::array<char, 64>> sync_encryption_key(util::none);
        if (!m_config.encryption_key.empty()) {
            sync_encryption_key = std::array<char, 64>();
            std::copy_n(m_config.encryption_key.begin(), 64, sync_encryption_key->begin());
        }

        m_config.sync_config = std::make_shared<SyncConfig>(SyncConfig{
            user, realm_url, SyncSessionStopPolicy::Immediately, std::move(bind_handler), std::move(error_handler),
            nullptr, sync_encryption_key, sync_client_validate_ssl, ssl_trust_certificate_path});

```

File: realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/EncryptedSynchronizedRealmTests.java [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+
+        // STEP 1: open a synced Realm using a local encryption key
+        String username = UUID.randomUUID().toString();
+        String password = "password";
+        SyncUser user = SyncUser.login(SyncCredentials.usernamePassword(username, password, true), Constants.AUTH_URL);
+
+        final byte[] randomKey = TestHelper.getRandomKey();
+
+        SyncConfiguration configWithEncryption = configurationFactory.createSyncConfigurationBuilder(user, Constants.USER_REALM)
+                .modules(new StringOnlyModule())
```
Comment (beeender) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/EncryptedSynchronizedRealmTests.java]: Won't `schemas(StringOnly.class)` work? uh... it is not available for `SyncConfiguration.Builder` ... shall we add it?

Comment (cmelchior) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/EncryptedSynchronizedRealmTests.java]: It should be there, but it is package protected so only works if the test is in `io.realm`

**Comment by cmelchior** [comment]
> What will happen if an app of old version is already using encryptionKey() and the app is updated to fixed version? Existing Realm database file will be encrypted automatically? Just an exception will be thrown?

This is actually kinda problematic. I don't think we are able to do it automatically because we cannot detect if a file is encrypted with the wrong key or not encrypted at all. Also, I think that  `writeEncryptedCopyTo` will write a compacted version, which means the transaction log is lost, making it un-syncable.

Only viable solution would be to make sure that all changes are uploaded, then delete the file and re-download everything again 😢 . We don't have a `Session.uploadAllServerChanges()` either yet.

IMO, we should just merge this as is, because this is really a bug that should be fixed ASAP

File: CHANGELOG.md [hunk 3]
```diff
@@ -7,6 +7,7 @@
 ### Bug Fixes
 
 * [ObjectServer] Fixed a crash when an authentication error happend (#4726).
+* [ObjectServer] Enable encryption with Sync (#4561).
```
Comment (kneth) [inline CHANGELOG.md]: "Enabled"?

File: realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/EncryptedSynchronizedRealmTests.java [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+public class EncryptedSynchronizedRealmTests extends BaseIntegrationTest {
+    @Rule
+    public Timeout globalTimeout = Timeout.seconds(10);
+
+    @Rule
+    public final TestSyncConfigurationFactory configurationFactory = new TestSyncConfigurationFactory();
+
+    @Before
+    public void before() {
+        // This will set the 'm_metadata_manager' in 'sync_manager.cpp' to be 'null'
```
Comment (kneth) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/EncryptedSynchronizedRealmTests.java]: Backquote like in line 47?

File: realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/EncryptedSynchronizedRealmTests.java (hunk 2 above)
Comment (nhachicha) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/EncryptedSynchronizedRealmTests.java]: ☝🏼

Comment (beeender) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/EncryptedSynchronizedRealmTests.java]: OK, that is good enough

File: realm/realm-library/src/androidTestObjectServer/java/io/realm/util/SyncTestUtils.java [hunk 5]
```diff
@@ -116,4 +129,14 @@ public static AuthenticateResponse createRefreshResponse() {
     public static AuthenticateResponse createErrorResponse(ErrorCode code) {
         return AuthenticateResponse.from(new ObjectServerError(code, "dummy"));
     }
+
+    public static void resetSyncMetadata() {
+        try {
+            SYNC_MANAGER_RESET_METHOD.invoke(null);
+        } catch (InvocationTargetException e) {
```
Comment (nhachicha) [inline realm/realm-library/src/androidTestObjectServer/java/io/realm/util/SyncTestUtils.java]: it requires API 19 unfortunately 
screen shot 2017-06-02 at 12 51 03

File: realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/EncryptedSynchronizedRealmTests.java [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+
+        // STEP 2: make sure the changes gets to the server
+        SystemClock.sleep(TimeUnit.SECONDS.toMillis(2));  // FIXME: Replace with Sync Progress Notifications once available.
+        realm.close(); // Realm is not deleted, just closed
+        user.logout();
+
+        // STEP 3: try to open again the Realm without the encryption key should fail
+        user = SyncUser.login(SyncCredentials.usernamePassword(username, password, false), Constants.AUTH_URL);
+        SyncConfiguration configWithoutEncryption = configurationFactory.createSyncConfigurationBuilder(user, Constants.USER_REALM)
+                .modules(new StringOnlyModule())
```
Comment (nhachicha) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/EncryptedSynchronizedRealmTests.java]: I want to keep the test a bit structured under `objectserver`

**Pull request merged**